storage/innobase/trx/trx0purge.cc

```diff
@@ -788,12 +788,11 @@ buf_block_t *purge_sys_t::get_page(page_id_t id)
 
   if (UNIV_LIKELY(undo_page != nullptr))
   {
-    undo_page->fix();
-    mtr.commit();
-    return undo_page;
+    undo_page->page.lock.s_unlock();
+    return undo_page; // batch_cleanup() will unfix()
   }
 
-  mtr.commit();
+  ut_ad(mtr.is_empty());
```

Comment:
The motivation of this change is to avoid a pair of atomic increment and decrement; `mtr.commit()` would execute `undo_page->unfix()`, which we want to avoid. This is a read-only operation, so it has nothing to do with enabling or disabling the write-ahead log.

It could be better to remove the `mtr_t*` parameter from `buf_page_get_gen()` and add a wrapper that would register the operation in a `mtr_t::m_log`. We do not really need any mini-transaction here.

Reply:
I realized that we can also avoid the `s_unlock()` by using the `RW_NO_LATCH` mode. If the page is read-fixed, then `buf_page_get_low()` will wait for a page latch. If the page is already in the buffer pool, all we need is to buffer-fix it.

Reply:
I will implement `buf_pool_t::undo_page_fix()` for this special case. It will also be beneficial in `purge_sys_t::view_guard::get()`, although there we must register the buffer-fixed block in the provided `mtr`.